### PR TITLE
Task/ctv 3223 ima videojs sample

### PR DESCRIPTION
**Pull request by rblaa**
### JIRA
https://truextech.atlassian.net/browse/CTV-3223

### Description
* Add a videjs version of this ref app, that uses the videojs component and IMA plugin to get a working player with ad playback.
* NOTE: simple-video-controller.js is pretty much the old video-controller.js file that was previously reviewed.
* videojs-controller.js is a copy of that, subsequently changed to migrate to the videojs ima plugin SDK approach instead.

### Testing
* Run app locally, run it [hosted via a test branch](https://s3.amazonaws.com/ctv.truex.com/web/ref-app-IMA-CSAI/test/videojs-demo.html).

### Commit Message Subject
CTV-3223: HTML5: create a Google IMA ref app that works with client-side ads

### Related PRs
n/a

### Checks
- [ ] Includes unit test coverage _(if applicable)_
- [x] Includes functional test coverage _(at feature-level, if applicable)_
- [x] Proposed commit messages follow [team conventions](https://github.com/socialvibe/adlabs-wiki/blob/develop/git/README.md#commits)

### CODE REVIEW

**Comment by kylelamx** [comment]
Are we going to maintain 2 branches? one for videojs, one with the default player?

**Comment by rblaa** [comment]
> I was under the impression that we are going to maintain 2 branches? one for videojs, one with the default player?
> I guess.. we don't have to?

No, I put both apps in the same branch, the web pack build will expose both the simple and videojs demo at:
https://ctv.truex.com/web/ref-app-IMA-CSAI/develop/index.html
https://ctv.truex.com/web/ref-app-IMA-CSAI/develop/videojs-demo.html

**Comment by rblaa** [comment]
You can try the test versions now as well:
https://ctv.truex.com/web/ref-app-IMA-CSAI/test/index.html
https://ctv.truex.com/web/ref-app-IMA-CSAI/test/videojs-demo.html

**Comment by rblaa** [comment]
@kylelamx I found this way easier to maintain, the simple-video-controller.js and videjs-controller.js files being the actual different implementations.

**Review by kylelamx — COMMENTED** [review]
Looks great to me. But Ace would be a better judge here. 
I wish there are more comments, and markers around true[X] codes like... resumeAdPlayback, for example. 
I would have no idea that this function means resuming the rest of the ad break after a true[X] opt-out, for example.

File: src/main.js [hunk 1]
```diff
@@ -27,13 +29,17 @@ export function main(videoControllerClass) {
 
     const spinner = new LoadingSpinner();
 
+    const currentVideoStream = videoStreams[0];
+
+    // Randomize the stream id and current user id for this session, to work around ad usage capping.
```
[the remark this answers is not in the document]
Comment (rblaa) [reply src/main.js:34]: Will do

**Comment by rblaa** [comment]
> Looks great to me. But Ace would be a better judge here.
> I wish there are more comments, and markers around true[X] codes like... resumeAdPlayback, for example.
> I would have no idea that this function means resuming the rest of the ad break after a true[X] opt-out, for example.

Would resumeAdBreakPlayback() be a better name?

File: src/videojs/videojs-controller.js [hunk 2]
```diff
@@ -137,73 +138,61 @@ export class BaseVideoController {
...
 
-        // Listen and respond to ads loaded and error events.
-        this.adsLoader.addEventListener(
-            google.ima.AdsManagerLoadedEvent.Type.ADS_MANAGER_LOADED, this.onAdsManagerLoaded, false);
-        this.adsLoader.addEventListener(
-            google.ima.AdErrorEvent.Type.AD_ERROR, this.onAdError, false);
+        // Normal scenario is to request a VAST VMAP ad playlist via a url.
+        // For this demo application, we will use a canned xml response.
+        const imaOptions = {
+            // adTagUrl: 'https://pubads.g.doubleclick.net/gampad/ads?' +
```
Comment (addisonm) [inline src/videojs/videojs-controller.js:164]: Unless HTML is workng fine, suggest removing the vast files here and only leave vmap, or use other vmap samples.

File: README.md [hunk 3]
```diff
@@ -10,27 +10,27 @@ For a more detailed true[X] integration guide, please refer to the [CTV Web Inte
...
 
 # Build/Develop/Deploy
 
 To begin development, run the standard `npm install` to download the project's dependencies.
 
-To deploy in general, one makes a deployable version in the `./dist` folder via `npm run build` and then hosts those contents somewhere appropriate. On then ensures the various platform installer configurations refer to that url.
-
-The hosted copy of this reference app is currently hosted at [https://ctv.truex.com/web/ref-app-IMA-CSAI/master/index.html](https://ctv.truex.com/web/ref-app-IMA-CSAI/master/index.html), which can be viewed directly in Chrome to review and debug the reference app generically.
+To deploy in general, one makes a deployable version in the `./dist` folder via `npm run build` and then hosts those contents somewhere appropriate. On then ensures the various platform installer configurations refer to that url. Again, [simple](https://ctv.truex.com/web/ref-app-IMA-CSAI/master/index.html) and [videojs-demo](https://ctv.truex.com/web/ref-app-IMA-CSAI/master/videojs-demo.html) hosted copies of the reference app are available for viewing in a browser, to review and debug the reference app generically.
```
Comment (addisonm) [inline README.md:33]: Typo in "On then ensures".  "One then ensures...." sounds like the intent.

**Review by addisonm — APPROVED** [review]
Some minor comments.  The code unfortunately feels rather repetitious if we needed to do maintenance or updates, but also part of the design of the two approaches.  I imagine that these apps will get updated irregularly so should be okay.

Looks fine otherwise.

**Comment by addisonm** [comment]
Unrelated.  I noticed that git now seems to default/force me to submit my a review to bulk send my comments.  Not sure if anyone else has seen similar.  Should be okay normally, but might leave some gaps or "not-submitted reviews"

**Comment by rblaa** [comment]
> Unrelated. I noticed that git now seems to default/force me to submit my a review to bulk send my comments. Not sure if anyone else has seen similar. Should be okay normally, but might leave some gaps or "not-submitted reviews"

This is actually a feature not a bug. I.e. people do not see your comments until you approve or reject things.

I have seen, e.g. in stash, where people start responding to my review comments before I have had the chance to get my "story" straight, e.g. to correct original inaccurate comments, etc.

File: src/videojs/videojs-controller.js (hunk 2 above)
Comment (rblaa) [reply src/videojs/videojs-controller.js:164]: Both examples were working for HTML5, is I left them in to allow exploration.
But I can remove to minimize distractions.

**Comment by rblaa** [comment]
> The code unfortunately feels rather repetitious if we needed to do maintenance or updates, but also part of the design of the two approaches. I imagine that these apps will get updated irregularly so should be okay.

What happen to me was that the effort to maintain things to have refactored code between the two controllers was really slowing down the development effort.

Perhaps I can do it now, but that's not the point of this demo, which is to present code clearly "all at once", so either the simple IMA SDK approach vs the videojs plugin approach

**Pull request merged**